File: android/src/main/java/com/rnvisioncameramlkit/stacked/StackedTextDetector.kt
package com.rnvisioncameramlkit.stacked

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

object StackedTextDetector {

    data class Box(val left: Int, val top: Int, val right: Int, val bottom: Int) {
        val width: Int get() = right - left
        val height: Int get() = bottom - top
        val centerX: Float get() = (left + right) / 2f
        val centerY: Float get() = (top + bottom) / 2f
    }

    data class Column(val bounds: Box, val glyphs: List<Box>)

    class Refined(val glyphs: List<Box>, val boxed: BooleanArray, val score: Float)

    private val BRIGHT_THRESHOLDS = intArrayOf(180, 210, 150)
    private const val DARK_THRESHOLD = 90
    private const val DARK_PASS = 3
    private const val PASS_COUNT = 4
    private const val MAX_CHROMA = 85
    private const val MAX_FOREGROUND_FRACTION = 0.40f
    private const val MIN_HEIGHT_FRACTION = 0.006f
    private const val MAX_HEIGHT_FRACTION = 0.10f
    private const val MIN_WIDTH_FRACTION = 0.004f
    private const val MAX_WIDTH_FRACTION = 0.07f
    private const val MIN_ASPECT = 0.8f
    private const val MAX_ASPECT = 6f
    private const val MIN_FILL = 0.2f
    const val MIN_GLYPHS_PER_COLUMN = 8
    private const val MAX_VERTICAL_GAP_IN_HEIGHTS = 3f
    private const val DUPLICATE_IOU = 0.55f

    const val CONTAINER_NUMBER_LENGTH = 11
    private const val MERGED_HEIGHT_RATIO = 1.7f
    private const val BOXED_WIDTH_RATIO = 1.35f
    private const val BOXED_HEIGHT_RATIO = 1.12f
    private const val OUTLIER_MIN_RATIO = 0.6f
    private const val OUTLIER_MAX_RATIO = 1.6f
    private const val OFF_CENTRE_IN_WIDTHS = 0.5f
    private const val SCORE_PER_MISSING_GLYPH = 0.5f
    private const val SCORE_OFF_CENTRE_WEIGHT = 2f

    class Workspace {
        internal var low = ByteArray(0)
        internal var high = ByteArray(0)
        internal val lowHistogram = IntArray(256)
        internal val highHistogram = IntArray(256)
        internal var rowStart = IntArray(0)
        internal var runStart = IntArray(4096)
        internal var runEnd = IntArray(4096)
        internal var runRow = IntArray(4096)
        internal var parent = IntArray(4096)
        internal var boxLeft = IntArray(4096)
        internal var boxTop = IntArray(4096)
        internal var boxRight = IntArray(4096)
        internal var boxBottom = IntArray(4096)
        internal var area = IntArray(4096)
        var passesRun = 0
            internal set

        internal fun ensurePixels(count: Int) {
            if (low.size < count) low = ByteArray(count)
            if (high.size < count) high = ByteArray(count)
        }

        internal fun ensureRows(height: Int) {
            if (rowStart.size < height + 1) rowStart = IntArray(height + 1)
        }

        internal fun growRuns() {
            val size = runStart.size * 2
            runStart = runStart.copyOf(size)
            runEnd = runEnd.copyOf(size)
            runRow = runRow.copyOf(size)
            parent = parent.copyOf(size)
            boxLeft = IntArray(size)
            boxTop = IntArray(size)
            boxRight = IntArray(size)
            boxBottom = IntArray(size)
            area = IntArray(size)
        }
    }

    @JvmOverloads
    fun detect(
        pixels: IntArray,
        width: Int,
        height: Int,
        workspace: Workspace = Workspace()
    ): List<Column> {
        if (width < 2 || height < 2) return emptyList()
        val count = width * height
        require(pixels.size >= count)
        workspace.ensurePixels(count)

        val low = workspace.low
        val high = workspace.high
        val lowHistogram = workspace.lowHistogram
        val highHistogram = workspace.highHistogram
        lowHistogram.fill(0)
        highHistogram.fill(0)
        for (i in 0 until count) {
            val pixel = pixels[i]
            val r = (pixel shr 16) and 0xFF
            val g = (pixel shr 8) and 0xFF
            val b = pixel and 0xFF
            val lo = min(r, min(g, b))
            val hi = max(r, max(g, b))
            low[i] = lo.toByte()
            high[i] = hi.toByte()
            if (hi - lo < MAX_CHROMA) lowHistogram[lo]++
            highHistogram[hi]++
        }
        return detect(workspace, low, high, width, height)
    }

    @JvmOverloads
    fun detect(
        luma: ByteArray,
        width: Int,
        height: Int,
        workspace: Workspace = Workspace()
    ): List<Column> {
        if (width < 2 || height < 2) return emptyList()
        val count = width * height
        require(luma.size >= count)

        val histogram = workspace.highHistogram
        histogram.fill(0)
        for (i in 0 until count) histogram[luma[i].toInt() and 0xFF]++
        histogram.copyInto(workspace.lowHistogram)
        return detect(workspace, luma, luma, width, height)
    }

    fun refine(column: Column): Refined = refine(column.glyphs)

    private fun detect(workspace: Workspace, low: ByteArray, high: ByteArray, width: Int, height: Int): List<Column> {
        workspace.ensureRows(height)
        workspace.passesRun = 0
        val found = ArrayList<Column>()
        for (pass in 0 until PASS_COUNT) collectColumns(workspace, low, high, width, height, pass, found)
        return deduplicate(found)
    }

    private fun collectColumns(
        workspace: Workspace,
        low: ByteArray,
        high: ByteArray,
        width: Int,
        height: Int,
        pass: Int,
        into: MutableList<Column>
    ) {
        val bright = pass != DARK_PASS
        val threshold = if (bright) BRIGHT_THRESHOLDS[pass] else DARK_THRESHOLD
        val count = width * height

        var foreground = 0
        if (bright) {
            for (v in threshold..255) foreground += workspace.lowHistogram[v]
        } else {
            for (v in 0 until threshold) foreground += workspace.highHistogram[v]
        }
        if (foreground == 0 || foreground > count * MAX_FOREGROUND_FRACTION) return

        workspace.passesRun++
        val glyphs = labelGlyphs(workspace, low, high, width, height, bright, threshold)
        if (glyphs.size < MIN_GLYPHS_PER_COLUMN) return
        into += cluster(glyphs)
    }

    internal fun labelGlyphs(
        workspace: Workspace,
        low: ByteArray,
        high: ByteArray,
        width: Int,
        height: Int,
        bright: Boolean,
        threshold: Int
    ): List<Box> {
        workspace.ensureRows(height)
        val rowStart = workspace.rowStart
        var runStart = workspace.runStart
        var runEnd = workspace.runEnd
        var runRow = workspace.runRow
        var parent = workspace.parent
        var runs = 0

        for (y in 0 until height) {
            rowStart[y] = runs
            val previousFrom = if (y > 0) rowStart[y - 1] else 0
            val previousTo = runs
            val rowOffset = y * width
            var j = previousFrom
            var x = 0
            while (x < width) {
                while (x < width && !isOn(low, high, rowOffset + x, bright, threshold)) x++
                if (x >= width) break
                val start = x
                while (x < width && isOn(low, high, rowOffset + x, bright, threshold)) x++

                if (runs == runStart.size) {
                    workspace.growRuns()
                    runStart = workspace.runStart
                    runEnd = workspace.runEnd
                    runRow = workspace.runRow
                    parent = workspace.parent
                }
                val run = runs++
                runStart[run] = start
                runEnd[run] = x
                runRow[run] = y
                parent[run] = run

                while (j < previousTo && runEnd[j] <= start) j++
                var k = j
                while (k < previousTo && runStart[k] < x) {
                    union(parent, run, k)
                    k++
                }
            }
        }
        rowStart[height] = runs

        val boxLeft = workspace.boxLeft
        val boxTop = workspace.boxTop
        val boxRight = workspace.boxRight
        val boxBottom = workspace.boxBottom
        val area = workspace.area
        for (i in 0 until runs) {
            if (parent[i] != i) continue
            boxLeft[i] = runStart[i]
            boxRight[i] = runEnd[i]
            boxTop[i] = runRow[i]
            boxBottom[i] = runRow[i] + 1
            area[i] = runEnd[i] - runStart[i]
        }
        for (i in 0 until runs) {
            if (parent[i] == i) continue
            val root = find(parent, i)
            if (runStart[i] < boxLeft[root]) boxLeft[root] = runStart[i]
            if (runEnd[i] > boxRight[root]) boxRight[root] = runEnd[i]
            if (runRow[i] + 1 > boxBottom[root]) boxBottom[root] = runRow[i] + 1
            area[root] += runEnd[i] - runStart[i]
        }

        val longSide = max(width, height)
        val minHeight = MIN_HEIGHT_FRACTION * longSide
        val maxHeight = MAX_HEIGHT_FRACTION * longSide
        val minWidth = MIN_WIDTH_FRACTION * longSide
        val maxWidth = MAX_WIDTH_FRACTION * longSide
        val glyphs = ArrayList<Box>()
        for (i in 0 until runs) {
            if (parent[i] != i) continue
            val boxWidth = boxRight[i] - boxLeft[i]
            val boxHeight = boxBottom[i] - boxTop[i]
            if (boxWidth < minWidth || boxWidth > maxWidth) continue
            if (boxHeight < minHeight || boxHeight > maxHeight) continue
            val aspect = boxHeight.toFloat() / boxWidth
            if (aspect < MIN_ASPECT || aspect > MAX_ASPECT) continue
            if (area[i].toFloat() / (boxWidth * boxHeight) < MIN_FILL) continue
            glyphs += Box(boxLeft[i], boxTop[i], boxRight[i], boxBottom[i])
        }
        return glyphs
    }

    @Suppress("NOTHING_TO_INLINE")
    private inline fun isOn(low: ByteArray, high: ByteArray, index: Int, bright: Boolean, threshold: Int): Boolean {
        val lo = low[index].toInt() and 0xFF
        val hi = high[index].toInt() and 0xFF
        return if (bright) lo >= threshold && hi - lo < MAX_CHROMA else hi < threshold
    }

    private fun find(parent: IntArray, index: Int): Int {
        var i = index
        while (parent[i] != i) {
            parent[i] = parent[parent[i]]
            i = parent[i]
        }
        return i
    }

    private fun union(parent: IntArray, a: Int, b: Int) {
        val rootA = find(parent, a)
        val rootB = find(parent, b)
        if (rootA == rootB) return
        if (rootA < rootB) parent[rootB] = rootA else parent[rootA] = rootB
    }

    private fun cluster(glyphs: List<Box>): List<Column> {
        val ordered = glyphs.sortedWith(compareBy({ it.top }, { it.left }))
        val open = mutableListOf<MutableList<Box>>()

        for (glyph in ordered) {
            var best: MutableList<Box>? = null
            var bestDistance = Float.MAX_VALUE
            for (column in open) {
                val last = column.last()
                val distance = abs(glyph.centerX - last.centerX)
                if (distance > max(glyph.width, last.width)) continue
                if (glyph.top - last.bottom > MAX_VERTICAL_GAP_IN_HEIGHTS * last.height) continue
                if (distance < bestDistance) {
                    bestDistance = distance
                    best = column
                }
            }
            if (best != null) best += glyph else open += mutableListOf(glyph)
        }

        return open.filter { it.size >= MIN_GLYPHS_PER_COLUMN }.map { column ->
            Column(boundsOf(column), column.toList())
        }
    }

    private fun boundsOf(glyphs: List<Box>): Box {
        var left = Int.MAX_VALUE
        var top = Int.MAX_VALUE
        var right = Int.MIN_VALUE
        var bottom = Int.MIN_VALUE
        for (glyph in glyphs) {
            if (glyph.left < left) left = glyph.left
            if (glyph.top < top) top = glyph.top
            if (glyph.right > right) right = glyph.right
            if (glyph.bottom > bottom) bottom = glyph.bottom
        }
        return Box(left, top, right, bottom)
    }

    private fun deduplicate(columns: List<Column>): List<Column> {
        val kept = mutableListOf<Column>()
        val ranked = columns.sortedWith(
            compareByDescending<Column> { it.glyphs.size }
                .thenBy { it.bounds.top }
                .thenBy { it.bounds.left }
        )
        for (column in ranked) {
            if (kept.none { intersectionOverUnion(it.bounds, column.bounds) > DUPLICATE_IOU }) kept += column
        }
        return kept
    }

    private fun intersectionOverUnion(a: Box, b: Box): Float {
        val width = max(0, min(a.right, b.right) - max(a.left, b.left))
        val height = max(0, min(a.bottom, b.bottom) - max(a.top, b.top))
        val intersection = (width * height).toFloat()
        val union = a.width.toFloat() * a.height + b.width.toFloat() * b.height - intersection
        return if (union <= 0f) 0f else intersection / union
    }

    private fun refine(detected: List<Box>): Refined {
        val medianWidth = median(detected) { it.width }
        val medianHeight = median(detected) { it.height }

        val split = ArrayList<Box>(detected.size + 2)
        for (g in detected) {
            val parts = (g.height.toFloat() / medianHeight).roundToInt()
            if (parts < 2 || g.height <= medianHeight * MERGED_HEIGHT_RATIO) {
                split += g
                continue
            }
            val step = g.height.toFloat() / parts
            for (k in 0 until parts) {
                split += Box(g.left, g.top + (k * step).toInt(), g.right, g.top + ((k + 1) * step).toInt())
            }
        }

        val boxed = BooleanArray(split.size)
        val glyphs = ArrayList<Box>(split.size)
        for (i in split.indices) {
            val g = split[i]
            if (g.width > medianWidth * BOXED_WIDTH_RATIO && g.height > medianHeight * BOXED_HEIGHT_RATIO) {
                boxed[i] = true
                val dx = (g.width - medianWidth) / 2
                val dy = (g.height - medianHeight) / 2
                glyphs += Box(g.left + dx, g.top + dy, g.right - dx, g.bottom - dy)
            } else {
                glyphs += g
            }
        }

        var from = 0
        var to = glyphs.size
        while (to - from > CONTAINER_NUMBER_LENGTH && isOutlier(glyphs[from], medianWidth, medianHeight)) from++
        while (to - from > CONTAINER_NUMBER_LENGTH && isOutlier(glyphs[to - 1], medianWidth, medianHeight)) to--
        val trimmed = if (from == 0 && to == glyphs.size) glyphs else glyphs.subList(from, to)
        val trimmedBoxed = if (from == 0 && to == glyphs.size) boxed else boxed.copyOfRange(from, to)

        return Refined(trimmed, trimmedBoxed, score(trimmed))
    }

    private fun isOutlier(glyph: Box, medianWidth: Int, medianHeight: Int): Boolean =
        glyph.width < medianWidth * OUTLIER_MIN_RATIO || glyph.width > medianWidth * OUTLIER_MAX_RATIO ||
            glyph.height < medianHeight * OUTLIER_MIN_RATIO || glyph.height > medianHeight * OUTLIER_MAX_RATIO

    internal fun score(glyphs: List<Box>): Float {
        if (glyphs.size < 2) return Float.MAX_VALUE
        val medianWidth = median(glyphs) { it.width }
        var offCentre = 0
        val medianCentre = medianFloat(glyphs) { it.centerX }
        for (g in glyphs) if (abs(g.centerX - medianCentre) > OFF_CENTRE_IN_WIDTHS * medianWidth) offCentre++
        return abs(glyphs.size - CONTAINER_NUMBER_LENGTH) * SCORE_PER_MISSING_GLYPH +
            pitchVariation(glyphs) +
            variation(glyphs) { it.height.toFloat() } +
            variation(glyphs) { it.width.toFloat() } +
            SCORE_OFF_CENTRE_WEIGHT * offCentre / glyphs.size
    }

    private fun pitchVariation(glyphs: List<Box>): Float {
        if (glyphs.size < 3) return 0f
        val pitches = FloatArray(glyphs.size - 1) { glyphs[it + 1].centerY - glyphs[it].centerY }
        return variation(pitches)
    }

    private inline fun variation(glyphs: List<Box>, value: (Box) -> Float): Float =
        variation(FloatArray(glyphs.size) { value(glyphs[it]) })

    private fun variation(values: FloatArray): Float {
        if (values.isEmpty()) return 0f
        var sum = 0f
        for (v in values) sum += v
        val mean = sum / values.size
        if (mean == 0f) return Float.MAX_VALUE
        var squares = 0f
        for (v in values) squares += (v - mean) * (v - mean)
        return sqrt(squares / values.size) / abs(mean)
    }

    private inline fun median(glyphs: List<Box>, value: (Box) -> Int): Int {
        val sorted = IntArray(glyphs.size) { value(glyphs[it]) }
        sorted.sort()
        return sorted[sorted.size / 2]
    }

    private inline fun medianFloat(glyphs: List<Box>, value: (Box) -> Float): Float {
        val sorted = FloatArray(glyphs.size) { value(glyphs[it]) }
        sorted.sort()
        return sorted[sorted.size / 2]
    }
}

File: ios/StackedTextDetector.swift
//
//  StackedTextDetector.swift
//  react-native-vision-camera-ml-kit
//

import Foundation

public enum StackedTextDetector {

    public struct Box: Equatable {
        public let left: Int
        public let top: Int
        public let right: Int
        public let bottom: Int

        public var width: Int { right - left }
        public var height: Int { bottom - top }
        public var centerX: Double { Double(left + right) / 2.0 }
        public var centerY: Double { Double(top + bottom) / 2.0 }
    }

    public struct Column: Equatable {
        public let bounds: Box
        public let glyphs: [Box]
    }

    public struct Refined {
        public let glyphs: [Box]
        public let boxed: [Bool]
        public let score: Double
    }

    private static let brightThresholds = [180, 210, 150]
    private static let darkThreshold = 90
    private static let darkPass = 3
    private static let passCount = 4
    private static let maxChroma = 85
    private static let maxForegroundFraction = 0.40
    private static let minHeightFraction = 0.006
    private static let maxHeightFraction = 0.10
    private static let minWidthFraction = 0.004
    private static let maxWidthFraction = 0.07
    private static let minAspect = 0.8
    private static let maxAspect = 6.0
    private static let minFill = 0.2
    public static let minGlyphsPerColumn = 8
    private static let maxVerticalGapInHeights = 3.0
    private static let duplicateIoU = 0.55

    public static let containerNumberLength = 11
    private static let mergedHeightRatio = 1.7
    private static let boxedWidthRatio = 1.35
    private static let boxedHeightRatio = 1.12
    private static let outlierMinRatio = 0.6
    private static let outlierMaxRatio = 1.6
    private static let offCentreInWidths = 0.5
    private static let scorePerMissingGlyph = 0.5
    private static let scoreOffCentreWeight = 2.0

    public final class Workspace {
        fileprivate var low = [UInt8]()
        fileprivate var high = [UInt8]()
        fileprivate var lowHistogram = [Int](repeating: 0, count: 256)
        fileprivate var highHistogram = [Int](repeating: 0, count: 256)
        fileprivate var rowStart = [Int]()
        fileprivate var runStart = [Int](repeating: 0, count: 4096)
        fileprivate var runEnd = [Int](repeating: 0, count: 4096)
        fileprivate var runRow = [Int](repeating: 0, count: 4096)
        fileprivate var parent = [Int](repeating: 0, count: 4096)
        fileprivate var boxLeft = [Int](repeating: 0, count: 4096)
        fileprivate var boxTop = [Int](repeating: 0, count: 4096)
        fileprivate var boxRight = [Int](repeating: 0, count: 4096)
        fileprivate var boxBottom = [Int](repeating: 0, count: 4096)
        fileprivate var area = [Int](repeating: 0, count: 4096)
        public private(set) var passesRun = 0

        public init() {}

        fileprivate func ensurePixels(_ count: Int) {
            if low.count < count { low = [UInt8](repeating: 0, count: count) }
            if high.count < count { high = [UInt8](repeating: 0, count: count) }
        }

        fileprivate func ensureRows(_ height: Int) {
            if rowStart.count < height + 1 { rowStart = [Int](repeating: 0, count: height + 1) }
        }

        fileprivate func growRuns() {
            let size = runStart.count * 2
            runStart += [Int](repeating: 0, count: size - runStart.count)
            runEnd += [Int](repeating: 0, count: size - runEnd.count)
            runRow += [Int](repeating: 0, count: size - runRow.count)
            parent += [Int](repeating: 0, count: size - parent.count)
            boxLeft = [Int](repeating: 0, count: size)
            boxTop = [Int](repeating: 0, count: size)
            boxRight = [Int](repeating: 0, count: size)
            boxBottom = [Int](repeating: 0, count: size)
            area = [Int](repeating: 0, count: size)
        }

        fileprivate func resetPasses() { passesRun = 0 }

        fileprivate func countPass() { passesRun += 1 }
    }

    public static func detect(
        rgba: [UInt8],
        width: Int,
        height: Int,
        workspace: Workspace = Workspace()
    ) -> [Column] {
        guard width >= 2, height >= 2 else { return [] }
        let count = width * height
        guard rgba.count >= count * 4 else { return [] }
        workspace.ensurePixels(count)

        for i in 0..<256 {
            workspace.lowHistogram[i] = 0
            workspace.highHistogram[i] = 0
        }
        for i in 0..<count {
            let base = i * 4
            let r = rgba[base]
            let g = rgba[base + 1]
            let b = rgba[base + 2]
            let lo = Swift.min(r, Swift.min(g, b))
            let hi = Swift.max(r, Swift.max(g, b))
            workspace.low[i] = lo
            workspace.high[i] = hi
            if Int(hi) - Int(lo) < maxChroma { workspace.lowHistogram[Int(lo)] += 1 }
            workspace.highHistogram[Int(hi)] += 1
        }
        return workspace.low.withUnsafeBufferPointer { low in
            workspace.high.withUnsafeBufferPointer { high in
                detect(workspace, low, high, width, height)
            }
        }
    }

    public static func detect(
        luma: [UInt8],
        width: Int,
        height: Int,
        workspace: Workspace = Workspace()
    ) -> [Column] {
        guard width >= 2, height >= 2 else { return [] }
        let count = width * height
        guard luma.count >= count else { return [] }

        for i in 0..<256 { workspace.highHistogram[i] = 0 }
        for i in 0..<count { workspace.highHistogram[Int(luma[i])] += 1 }
        workspace.lowHistogram = workspace.highHistogram
        return luma.withUnsafeBufferPointer { pixels in
            detect(workspace, pixels, pixels, width, height)
        }
    }

    public static func refine(_ column: Column) -> Refined {
        refine(column.glyphs)
    }

    private static func detect(
        _ workspace: Workspace,
        _ low: UnsafeBufferPointer<UInt8>,
        _ high: UnsafeBufferPointer<UInt8>,
        _ width: Int,
        _ height: Int
    ) -> [Column] {
        workspace.ensureRows(height)
        workspace.resetPasses()
        var found = [Column]()
        for pass in 0..<passCount {
            found += collectColumns(workspace, low, high, width, height, pass: pass)
        }
        return deduplicate(found)
    }

    private static func collectColumns(
        _ workspace: Workspace,
        _ low: UnsafeBufferPointer<UInt8>,
        _ high: UnsafeBufferPointer<UInt8>,
        _ width: Int,
        _ height: Int,
        pass: Int
    ) -> [Column] {
        let bright = pass != darkPass
        let threshold = bright ? brightThresholds[pass] : darkThreshold
        let count = width * height

        var foreground = 0
        if bright {
            for v in threshold...255 { foreground += workspace.lowHistogram[v] }
        } else {
            for v in 0..<threshold { foreground += workspace.highHistogram[v] }
        }
        if foreground == 0 || Double(foreground) > Double(count) * maxForegroundFraction { return [] }

        workspace.countPass()
        let glyphs = labelGlyphs(workspace, low, high, width, height, bright: bright, threshold: threshold)
        if glyphs.count < minGlyphsPerColumn { return [] }
        return cluster(glyphs)
    }

    static func labelGlyphs(
        _ workspace: Workspace,
        _ low: UnsafeBufferPointer<UInt8>,
        _ high: UnsafeBufferPointer<UInt8>,
        _ width: Int,
        _ height: Int,
        bright: Bool,
        threshold: Int
    ) -> [Box] {
        workspace.ensureRows(height)
        var runs = 0

        @inline(__always) func isOn(_ index: Int) -> Bool {
            let lo = Int(low[index])
            let hi = Int(high[index])
            return bright ? (lo >= threshold && hi - lo < maxChroma) : (hi < threshold)
        }

        for y in 0..<height {
            workspace.rowStart[y] = runs
            let previousFrom = y > 0 ? workspace.rowStart[y - 1] : 0
            let previousTo = runs
            let rowOffset = y * width
            var j = previousFrom
            var x = 0
            while x < width {
                while x < width && !isOn(rowOffset + x) { x += 1 }
                if x >= width { break }
                let start = x
                while x < width && isOn(rowOffset + x) { x += 1 }

                if runs == workspace.runStart.count { workspace.growRuns() }
                let run = runs
                runs += 1
                workspace.runStart[run] = start
                workspace.runEnd[run] = x
                workspace.runRow[run] = y
                workspace.parent[run] = run

                while j < previousTo && workspace.runEnd[j] <= start { j += 1 }
                var k = j
                while k < previousTo && workspace.runStart[k] < x {
                    union(&workspace.parent, run, k)
                    k += 1
                }
            }
        }
        workspace.rowStart[height] = runs

        for i in 0..<runs where workspace.parent[i] == i {
            workspace.boxLeft[i] = workspace.runStart[i]
            workspace.boxRight[i] = workspace.runEnd[i]
            workspace.boxTop[i] = workspace.runRow[i]
            workspace.boxBottom[i] = workspace.runRow[i] + 1
            workspace.area[i] = workspace.runEnd[i] - workspace.runStart[i]
        }
        for i in 0..<runs where workspace.parent[i] != i {
            let root = find(&workspace.parent, i)
            if workspace.runStart[i] < workspace.boxLeft[root] { workspace.boxLeft[root] = workspace.runStart[i] }
            if workspace.runEnd[i] > workspace.boxRight[root] { workspace.boxRight[root] = workspace.runEnd[i] }
            if workspace.runRow[i] + 1 > workspace.boxBottom[root] { workspace.boxBottom[root] = workspace.runRow[i] + 1 }
            workspace.area[root] += workspace.runEnd[i] - workspace.runStart[i]
        }

        let longSide = Double(Swift.max(width, height))
        let minHeight = minHeightFraction * longSide
        let maxHeight = maxHeightFraction * longSide
        let minWidth = minWidthFraction * longSide
        let maxWidth = maxWidthFraction * longSide
        var glyphs = [Box]()
        for i in 0..<runs where workspace.parent[i] == i {
            let boxWidth = Double(workspace.boxRight[i] - workspace.boxLeft[i])
            let boxHeight = Double(workspace.boxBottom[i] - workspace.boxTop[i])
            if boxWidth < minWidth || boxWidth > maxWidth { continue }
            if boxHeight < minHeight || boxHeight > maxHeight { continue }
            let aspect = boxHeight / boxWidth
            if aspect < minAspect || aspect > maxAspect { continue }
            if Double(workspace.area[i]) / (boxWidth * boxHeight) < minFill { continue }
            glyphs.append(Box(
                left: workspace.boxLeft[i],
                top: workspace.boxTop[i],
                right: workspace.boxRight[i],
                bottom: workspace.boxBottom[i]
            ))
        }
        return glyphs
    }

    private static func find(_ parent: inout [Int], _ index: Int) -> Int {
        var i = index
        while parent[i] != i {
            parent[i] = parent[parent[i]]
            i = parent[i]
        }
        return i
    }

    private static func union(_ parent: inout [Int], _ a: Int, _ b: Int) {
        let rootA = find(&parent, a)
        let rootB = find(&parent, b)
        if rootA == rootB { return }
        if rootA < rootB { parent[rootB] = rootA } else { parent[rootA] = rootB }
    }

    private static func cluster(_ glyphs: [Box]) -> [Column] {
        let ordered = glyphs.sorted { ($0.top, $0.left) < ($1.top, $1.left) }
        var open = [[Box]]()

        for glyph in ordered {
            var best: Int?
            var bestDistance = Double.greatestFiniteMagnitude
            for (index, column) in open.enumerated() {
                guard let last = column.last else { continue }
                let distance = abs(glyph.centerX - last.centerX)
                if distance > Double(Swift.max(glyph.width, last.width)) { continue }
                if Double(glyph.top - last.bottom) > maxVerticalGapInHeights * Double(last.height) { continue }
                if distance < bestDistance {
                    bestDistance = distance
                    best = index
                }
            }
            if let best = best { open[best].append(glyph) } else { open.append([glyph]) }
        }

        return open.filter { $0.count >= minGlyphsPerColumn }.map { Column(bounds: bounds(of: $0), glyphs: $0) }
    }

    private static func bounds(of glyphs: [Box]) -> Box {
        Box(
            left: glyphs.map { $0.left }.min()!,
            top: glyphs.map { $0.top }.min()!,
            right: glyphs.map { $0.right }.max()!,
            bottom: glyphs.map { $0.bottom }.max()!
        )
    }

    private static func deduplicate(_ columns: [Column]) -> [Column] {
        let ranked = columns.sorted { a, b in
            if a.glyphs.count != b.glyphs.count { return a.glyphs.count > b.glyphs.count }
            if a.bounds.top != b.bounds.top { return a.bounds.top < b.bounds.top }
            return a.bounds.left < b.bounds.left
        }
        var kept = [Column]()
        for column in ranked {
            if !kept.contains(where: { intersectionOverUnion($0.bounds, column.bounds) > duplicateIoU }) {
                kept.append(column)
            }
        }
        return kept
    }

    private static func intersectionOverUnion(_ a: Box, _ b: Box) -> Double {
        let width = Swift.max(0, Swift.min(a.right, b.right) - Swift.max(a.left, b.left))
        let height = Swift.max(0, Swift.min(a.bottom, b.bottom) - Swift.max(a.top, b.top))
        let intersection = Double(width * height)
        let union = Double(a.width * a.height) + Double(b.width * b.height) - intersection
        return union <= 0 ? 0 : intersection / union
    }

    private static func refine(_ detected: [Box]) -> Refined {
        let medianWidth = median(detected.map { $0.width })
        let medianHeight = median(detected.map { $0.height })

        var split = [Box]()
        for g in detected {
            let parts = Int((Double(g.height) / Double(medianHeight)).rounded())
            if parts < 2 || Double(g.height) <= Double(medianHeight) * mergedHeightRatio {
                split.append(g)
                continue
            }
            let step = Double(g.height) / Double(parts)
            for k in 0..<parts {
                split.append(Box(
                    left: g.left,
                    top: g.top + Int(Double(k) * step),
                    right: g.right,
                    bottom: g.top + Int(Double(k + 1) * step)
                ))
            }
        }

        var boxed = [Bool](repeating: false, count: split.count)
        var glyphs = [Box]()
        glyphs.reserveCapacity(split.count)
        for (i, g) in split.enumerated() {
            if Double(g.width) > Double(medianWidth) * boxedWidthRatio,
               Double(g.height) > Double(medianHeight) * boxedHeightRatio {
                boxed[i] = true
                let dx = (g.width - medianWidth) / 2
                let dy = (g.height - medianHeight) / 2
                glyphs.append(Box(left: g.left + dx, top: g.top + dy, right: g.right - dx, bottom: g.bottom - dy))
            } else {
                glyphs.append(g)
            }
        }

        var from = 0
        var to = glyphs.count
        while to - from > containerNumberLength && isOutlier(glyphs[from], medianWidth, medianHeight) { from += 1 }
        while to - from > containerNumberLength && isOutlier(glyphs[to - 1], medianWidth, medianHeight) { to -= 1 }
        let trimmed = Array(glyphs[from..<to])
        return Refined(glyphs: trimmed, boxed: Array(boxed[from..<to]), score: score(trimmed))
    }

    private static func isOutlier(_ glyph: Box, _ medianWidth: Int, _ medianHeight: Int) -> Bool {
        let width = Double(glyph.width)
        let height = Double(glyph.height)
        return width < Double(medianWidth) * outlierMinRatio || width > Double(medianWidth) * outlierMaxRatio ||
            height < Double(medianHeight) * outlierMinRatio || height > Double(medianHeight) * outlierMaxRatio
    }

    static func score(_ glyphs: [Box]) -> Double {
        if glyphs.count < 2 { return Double.greatestFiniteMagnitude }
        let medianWidth = Double(median(glyphs.map { $0.width }))
        let medianCentre = median(glyphs.map { $0.centerX })
        let offCentre = glyphs.filter { abs($0.centerX - medianCentre) > offCentreInWidths * medianWidth }.count
        return Double(abs(glyphs.count - containerNumberLength)) * scorePerMissingGlyph +
            pitchVariation(glyphs) +
            variation(glyphs.map { Double($0.height) }) +
            variation(glyphs.map { Double($0.width) }) +
            scoreOffCentreWeight * Double(offCentre) / Double(glyphs.count)
    }

    private static func pitchVariation(_ glyphs: [Box]) -> Double {
        if glyphs.count < 3 { return 0 }
        var pitches = [Double]()
        for i in 1..<glyphs.count { pitches.append(glyphs[i].centerY - glyphs[i - 1].centerY) }
        return variation(pitches)
    }

    private static func variation(_ values: [Double]) -> Double {
        if values.isEmpty { return 0 }
        let mean = values.reduce(0, +) / Double(values.count)
        if mean == 0 { return Double.greatestFiniteMagnitude }
        let squares = values.reduce(0) { $0 + ($1 - mean) * ($1 - mean) }
        return (squares / Double(values.count)).squareRoot() / abs(mean)
    }

    private static func median<T: Comparable>(_ values: [T]) -> T {
        let sorted = values.sorted()
        return sorted[sorted.count / 2]
    }
}
